[Ready] Punching/Grabbing Rebalance

PKPenguin321 opened the pull request:
:cl:
balance: Humans now have a punch damage range of 1-10, up from 0-9
balance: The chance to miss a punch now increases the more stamina and brute damage you have
balance: The stun duration for a critical punch now scales with how much stamina/brute damage the victim has
balance: Punches now deal full stamina damage + half damage in brute (effectively 1.5x damage on all punches, not accounting for stamina regeneration)
balance: Kicking (punching a prone person) does full brute damage, never misses, and does 1.5x damage
balance: Grabs are now harder to break out of the more stamina/brute damage you have
balance: Failing to resist out of a grab will deal a bit of stamina damage to you
balance: In all instances above where both stamina and brute damage are considered, brute is weighed 1/2 as heavily as stamina (meaning stamina has more of an effect than brute when it comes to punch-stun duration, miss chance, etc)
/:cl:

Please see https://tgstation13.org/phpBB/viewtopic.php?f=9&t=22154#p485654 for the initial discussion thread

__Grab change goals:__
- Make choking worth going for without making it too useful in any scenario
- Make restraining weakened people with grabs more reliable for things like kicking out assistants non-lethally

__Punch change goals:__
- Make fistfights less "I got an RNG stunpunch, I win!"
- Give fistfights a nonlethal way of ending instead of one person inevitably going into crit (stamina stunning, stamina damage punches -> hard-to-break-free-of grab)
- Make punching play into the new grab changes by dealing stamina damage (see above point)
- Retain the lethality of fistfights should the participants deliberately take it to the death (1.5x damage on everything, potential for longer stuns, more raw damage when kicking)

PKPenguin321 committed:
initial commit

PKPenguin321 committed:
obligatory stupid mistake fixer commit
min -> max

actioninja commented:
Ehhh, could be better, but an improvement is an improvement.

knacker48 commented:
I like the idea of this but I'm concerned about the stamina damage. Like do punches deal a lot of stamina or very little? If its a lot then how many punches does it take before you're too slow to run away?

Nickkoranda commented:
>I like the idea of this but I'm concerned about the stamina damage. Like do punches deal a lot of stamina or very little? If its a lot then how many punches does it take before you're too slow to run away?

>balance: Humans now have a punch damage range of 1-10, up from 0-9

>balance: Punches now deal full stamina damage + half damage in brute (effectively 1.5x damage on all punches, not accounting for stamina regeneration)

The real question is how does this affect ethereals?

Shadowflame909 commented:
I like it!

But does this mean Boxing Gloves are now useless? Since their whole gimmick was "Punching but it does stamina damage/knock out"

PKPenguin321 commented:
Boxing gloves always do 100% stamina damage, can't disarm or grab at all, and can do KO punches on people with high stamina damage. They're still very different

ATH1909 commented:
> Boxing gloves always do 100% stamina damage, can't disarm or grab at all, and can do KO punches on people with high stamina damage. They're still very different

So a normal punch does stamina damage equal to that of a boxing glove punch (but also with some extra brute damage)? That seems kind of meh.

Maybe you should buff boxing gloves to 150% stamina damage, making them deal as much total damage as fists do (but just as pure stamina damage, rather than a mix of stamina and brute damage)?

necromanceranne commented:
How is this going to affect martial arts? Because they tend to change the damage on punches and I'm both concerned that either their damage is going to be converted into mostly stamina and half brute and they wind up with 150% more lethality than they did before, but that their ability to break out of grabs could fuck with combos. Carp in particular, since while it has a stun, pushing someone into stamina crit seems helpful, yet aggressive grabbing is how they open 95% of the time if it isn't a back kick.

Not that I have a problem with aggressive grab being harder to pull off.

AdamElTablawy commented:
i think this might make golems even more hellish as well

necromanceranne commented:
I'm even worried about gloves of the north star getting really wacky with any means to floor an opponent, because you are rapidly putting out 1.5x damage kicks. It's already pretty potent as it is.

Whoneedspacee commented:
gloves of the north star are already insta death with any stun anyways aren't they

francinum commented:
unarmed hits on stunned targets are now equivalent to a fire extinguisher. Even without the Gloves of the north star, that's really not a lot of hits to crit.

Whoneedspacee commented:
this combined with disarm wall stuns being less op would be neat and make sense tbh. 3 people curbstompimg you should fuck you up, especially unarmored.

PKPenguin321 commented:
Sorry for stale, been too busy to debug this. Will get to it eventually, hopefully this weekend?

ExcessiveUseOfCobblestone commented:
don't worry PRs I like get extra time before I call them stale

PKPenguin321 committed:
moves punch changes to correct file
makes them slightly more logical as well

PKPenguin321 committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into new-punches-and-grabs

# Conflicts:
#	code/modules/mob/living/living.dm

PKPenguin321 committed:
minor tweaks (see msg for details)
- fixed some defines i messed up when conflict fixing
- removed some unneeded typecasting and src calls
- bumped up stamina damage on failed grab resist so it's not as outpaced by stam regen

PKPenguin321 commented:
Looks good during local tests, will try to get a test merge

Mickyan commented:
Any particular reason why these calculations ignore damage types other than brute? 

As total health is the value for the mob's overall wellness (hence every damage type is factored in things such as slowdowns and crit thresholds), it seems counter-intuitive to only factor brute for this

PKPenguin321 commented:
I limited it to brute to synergize with the fact that (most) punches deal brute. I didn't want to prevent people who were damaged from other means to be able to put up a little bit of a fight. If it wasn't such a horrible snowflakey idea, I'd probably add a "fistfight" damage type that gets checked in place of brute.

adamkad1 commented:
hmmm, maybe you could add to muscle stimulant chem the ability to make you ignore stamina damage checks? right now it only makes you ignore slows

ExcessiveUseOfCobblestone commented:
wut

ExcessiveUseOfCobblestone commented:
Is this actually WIP or can it be reviewed

PKPenguin321 commented:
Been ready for review for a while

ExcessiveUseOfCobblestone commented:
cringes in labels

optimumtact commented on code/modules/mob/living/carbon/human/species.dm:
```diff
@@ -1224,12 +1229,19 @@ GLOBAL_LIST_EMPTY(roundstart_races)
 			target.dismembering_strike(user, affecting.body_zone)
-		target.apply_damage(damage, attack_type, affecting, armor_block)
+
+		if(atk_verb == "kick")//kicks deal 1.5x raw damage
```
wtf is this really how it works?

optimumtact replied:
please make these defines

optimumtact commented on code/modules/mob/living/living.dm:
```diff
@@ -719,13 +718,18 @@
 	if(pulledby.grab_state || resting)
 		var/altered_grab_state = pulledby.grab_state
-		if(resting) //If resting, resisting out of a grab is equivalent to 1 grab state higher
+		if(resting && pulledby.grab_state < 3) //If resting, resisting out of a grab is equivalent to 1 grab state higher. wont make the grab state exceed the normal max, however
```
grab states should have defines 100%

PKPenguin321 replied:
Turns out defines for this already existed, used em

optimumtact commented on code/modules/mob/living/living.dm:
```diff
@@ -719,13 +718,18 @@
+		if(resting && pulledby.grab_state < 3) //If resting, resisting out of a grab is equivalent to 1 grab state higher. wont make the grab state exceed the normal max, however
 			altered_grab_state++
-		if(prob(30/altered_grab_state))
+		var/resist_chance = 30
```
NO MAGIC NUMBERS

PKPenguin321 replied:
is `#define BASE_RESIST_CHANCE` ok?

ExcessiveUseOfCobblestone replied:
sure

necromanceranne commented:
Wait it doesn't factor burn?

So ethereals don't stand a chance in a fist fight?

PKPenguin321 commented:
Pretty much. Is that ok?

optimumtact commented:
sounds fine to me

PKPenguin321 committed:
better use of defines/logging

PKPenguin321 committed:
last second fix to stun punch threshold
forgot to raise it by 1 to account for raising the human punch damage max
also, log missed punches

PKPenguin321 commented:
Last two commits address reviews, add better logging to punches (logging missed punches/stun punches/kicks separately from regular punches), and also fix the stun punch threshold for humans since I forgot to increase it to accommodate the new punch damage maximum. PR should be good to merge

ExcessiveUseOfCobblestone commented on code/modules/mob/living/carbon/human/species.dm:
```diff
@@ -1224,13 +1230,22 @@ GLOBAL_LIST_EMPTY(roundstart_races)
 			target.visible_message("<span class='danger'>[user] has knocked  [target] down!</span>", \
 							"<span class='userdanger'>[user] has knocked [target] down!</span>", null, COMBAT_MESSAGE_RANGE)
-			target.apply_effect(80, EFFECT_KNOCKDOWN, armor_block)
+			var/knockdown_duration = 40 + (target.getStaminaLoss() + (target.getBruteLoss()/2))/1.25 //50 total damage = 40 base stun + 40 stun modifier = 80 stun duration, which is the old base duration
```
```suggestion
			var/knockdown_duration = 40 + (target.getStaminaLoss() + (target.getBruteLoss()*0.5))*0.8 //50 total damage = 40 base stun + 40 stun modifier = 80 stun duration, which is the old base duration
```

PKPenguin321 replied:
Why? Seems less readable

ExcessiveUseOfCobblestone replied:
`Do not divide when you can easily convert it to multiplication. (ie 4/2 should be done as 4*0.5)`

ExcessiveUseOfCobblestone commented on code/modules/mob/living/carbon/human/species.dm:
```diff
@@ -1224,13 +1230,22 @@ GLOBAL_LIST_EMPTY(roundstart_races)
+			target.apply_damage(damage*1.5, attack_type, affecting, armor_block)
+			log_combat(user, target, "kicked")
+		else//other attacks to 1/2 raw damage + full in stamina damage
+			target.apply_damage(damage/2, attack_type, affecting, armor_block)
```
```suggestion
			target.apply_damage(damage*0.5, attack_type, affecting, armor_block)
```

ExcessiveUseOfCobblestone commented on code/modules/mob/living/carbon/human/species.dm:
```diff
@@ -1204,13 +1204,19 @@ GLOBAL_LIST_EMPTY(roundstart_races)
+		if(atk_verb == ATTACK_EFFECT_KICK && user.dna.species.punchdamagelow != 0)//kicks never miss (provided your species deals more than 0 damage)
+			miss_chance = 0
+		else if(user.dna.species.punchdamagelow != 0)//don't divide by 0
+			miss_chance = min((user.dna.species.punchdamagehigh/user.dna.species.punchdamagelow) + user.getStaminaLoss() + (user.getBruteLoss()/2), 100) //old base chance for a miss + various damage. capped at 100 to prevent weirdness in prob()
```
```suggestion
			miss_chance = min((user.dna.species.punchdamagehigh/user.dna.species.punchdamagelow) + user.getStaminaLoss() + (user.getBruteLoss()*0.5), 100) //old base chance for a miss + various damage. capped at 100 to prevent weirdness in prob()
```

ExcessiveUseOfCobblestone commented on code/modules/mob/living/carbon/human/species.dm:
```diff
@@ -1204,13 +1204,19 @@ GLOBAL_LIST_EMPTY(roundstart_races)
+		var/miss_chance = 100//calculate the odds that a punch misses entirely. considers stamina and brute damage of the puncher. punches miss by default to prevent weird cases
+		if(atk_verb == ATTACK_EFFECT_KICK && user.dna.species.punchdamagelow != 0)//kicks never miss (provided your species deals more than 0 damage)
+			miss_chance = 0
+		else if(user.dna.species.punchdamagelow != 0)//don't divide by 0
```
```suggestion
		else if(user.dna.species.punchdamagelow)
```

ExcessiveUseOfCobblestone commented on code/modules/mob/living/carbon/human/species.dm:
```diff
@@ -1188,14 +1188,14 @@ GLOBAL_LIST_EMPTY(roundstart_races)
 
-		switch(atk_verb)
-			if("kick")
+		switch(atk_verb)//this code is really stupid but some genius apparently made "claw" and "slash" two attack types but also the same one so it's needed i guess
```
Not sure I understand the comment here but if they're the same do you think 1196 should be 

`if(ATTACK_EFFECT_SLASH || ATTACK_EFFECT_CLAW)`

??

PKPenguin321 replied:
Doesn't really fix the issue at hand, plus out of scope

ExcessiveUseOfCobblestone replied:
Regardless the claw attack should probably do the claw animation instead of punch, I'd consider that an oversight/bug that can be fixed within this PR.

PKPenguin321 replied:
It does not do a punch animation

ExcessiveUseOfCobblestone replied:
If `ATTACK_EFFECT_CLAW` or `"claw"` is used in `dna.species.attack_verb` it will use the punch attack animation according to this block of code.

PKPenguin321 replied:
Ah, I see what you mean.

ExcessiveUseOfCobblestone commented on code/modules/mob/living/carbon/human/species.dm:
```diff
@@ -1204,13 +1204,19 @@ GLOBAL_LIST_EMPTY(roundstart_races)
 
-		if(!damage || !affecting)
+		var/miss_chance = 100//calculate the odds that a punch misses entirely. considers stamina and brute damage of the puncher. punches miss by default to prevent weird cases
+		if(atk_verb == ATTACK_EFFECT_KICK && user.dna.species.punchdamagelow != 0)//kicks never miss (provided your species deals more than 0 damage)
```
```suggestion
		if(atk_verb == ATTACK_EFFECT_KICK && user.dna.species.punchdamagelow)//kicks never miss (provided your species deals more than 0 damage)
```

ExcessiveUseOfCobblestone replied:
You're checking punchdamagelow multiple times, what about putting a singular check in the front so you can ensure no bad math?

PKPenguin321 replied:
Will have to look at this later

ExcessiveUseOfCobblestone commented on code/modules/mob/living/living.dm:
```diff
@@ -719,13 +718,18 @@
 			altered_grab_state++
-		if(prob(30/altered_grab_state))
+		var/resist_chance = BASE_GRAB_RESIST_CHANCE // see defines/combat.dm
+		resist_chance = max(resist_chance/altered_grab_state-sqrt((getStaminaLoss()+getBruteLoss()/2)*(3-altered_grab_state)), 0) // https://i.imgur.com/6yAT90T.png for sample output values
```
rounding needed? Those r gross numbers.

PKPenguin321 replied:
Not really necessary. Maybe if the numbers were shown to the player but they aren't

ExcessiveUseOfCobblestone committed:
Update code/modules/mob/living/carbon/human/species.dm

Co-Authored-By: PKPenguin321 <pkpenguin321.git@gmail.com>

ExcessiveUseOfCobblestone committed:
Update code/modules/mob/living/carbon/human/species.dm

Co-Authored-By: PKPenguin321 <pkpenguin321.git@gmail.com>

ExcessiveUseOfCobblestone committed:
Update code/modules/mob/living/carbon/human/species.dm

Co-Authored-By: PKPenguin321 <pkpenguin321.git@gmail.com>

ExcessiveUseOfCobblestone committed:
Update code/modules/mob/living/carbon/human/species.dm

Co-Authored-By: PKPenguin321 <pkpenguin321.git@gmail.com>

necromanceranne commented:
You know, rather than leave ethereals with no ability to fight in of the games combat scenarios, you just give them brute damage on their punches until someone cares enough to change this to accommodate punching damage more broadly.

It was mostly a flavour thing anyway.

ExcessiveUseOfCobblestone commented:
so um how does this work with martial?

PKPenguin321 commented:
I mean ethreals aren't actually getting nerfed very much by this, they just don't confer brute for the tiny amount of extra miss chance it causes to the victim and their stun punches will be some variable amount between 25-50% shorter

PKPenguin321 commented:
Martial arts shouldn't be affected by this. Either martial arts override punches entirely or they just use them as part of a combo string and none of this code touches any of that at all

PKPenguin321 committed:
addresses review

ExcessiveUseOfCobblestone commented on code/modules/mob/living/carbon/human/species.dm:
```diff
@@ -1193,7 +1193,7 @@ GLOBAL_LIST_EMPTY(roundstart_races)
 			if(ATTACK_EFFECT_KICK)
 				user.do_attack_animation(target, ATTACK_EFFECT_KICK)
-			if(ATTACK_EFFECT_SLASH)
+			if(ATTACK_EFFECT_SLASH || ATTACK_EFFECT_CLAW)//smh
```
die

Mickyan commented:
I really like how this turned out, also gave a much needed indirect buff to slipping without making them an outright stun like it used to

I was going to suggest maybe increasing the stamina loss from a failed resist by factoring in the grab tier and/or existing damage, with the way stamina regen works spamming the resist button feels almost irrelevant.
I figure it could be more interesting if someone who's heavily damaged was more likely to cooperate or wait for the right opportunity instead of constantly struggling

ExcessiveUseOfCobblestone commented:
I'ma go ahead and merge since it's fine as is, any further tweaks can be added later.

Mickyan commented:
rude

(it's fine though)

PKPenguin321 commented:
>I was going to suggest maybe increasing the stamina loss from a failed resist by factoring in the grab tier and/or existing damage, with the way stamina regen works spamming the resist button feels almost irrelevant.

Would have done this had I seen it sooner
